### PR TITLE
Update .user.ini when setting upload size limit

### LINKED ISSUE #14748: File handling in admin: Also write to .user.ini
Hi,

it seems that the "file handling" input box in the admin backend to set a specific upload size is currently only writing to the .htaccess file: https://github.com/owncloud/core/blob/v8.0.0/lib/private/files.php#L217-L273

As newer OC versions are also shipping a .user.ini for other environments like fcgi the function probably could be extended to also write to this file.

### PR DESCRIPTION
**Pull request by RobinMcCorkell**
We try to change as many files as possible, so if we can write to `.user.ini` but not to `.htaccess` then we will (but the file contents will become inconsistent). Of course, whichever one actually affects the installation will be shown in the upload limit box in the admin settings, so there is no ambiguity in what happened, and the logs clearly state if any files could not be written to.

Not sure if this can be unit tested?

Fixes #14748

cc @PVince81 @LukasReschke @DeepDiver1975 @RealRancor @MorrisJobke

### CODE REVIEW

**Comment by mmattel** [comment]
:+1:

**Comment by karlitschek** [comment]
@DeepDiver1975 What do you think?

File: lib/private/files.php [hunk 1]
```diff
@@ -274,38 +275,63 @@ static function setUploadLimit($size) {
...
+					'Can\'t write upload limit to ' . $filename . '. Please check the file permissions',
+					OC_Log::WARN);
+				$success = false;
+				continue; // try to update as many files as possible
 			}
-			if ($hasReplaced === 0) {
-				$htaccess .= "\n" . $setting;
+
+			$content = '';
+			while (!feof($handle)) {
```
Comment (MorrisJobke) [inline lib/private/files.php]: Mayber use file_get_contents: https://secure.php.net/manual/de/function.file-get-contents.php

File: lib/private/files.php [hunk 2]
```diff
@@ -274,38 +275,63 @@ static function setUploadLimit($size) {
...
+					$content .= "\n" . $setting;
+				}
+			}
+
+			// write file back
+			ftruncate($handle, 0);
+			rewind($handle);
+			fwrite($handle, $content);
+
+			fclose($handle);
```
Comment (MorrisJobke) [inline lib/private/files.php]: Use file_put_contents instead of the 4 lines above: https://secure.php.net/manual/de/function.file-put-contents.php

**Comment by DeepDiver1975** [comment]
> Not sure if this can be unit tested?

I guess it is possible
- by passing in the root folder where files are to be read/written this can be tweaked for unit testing.

**Comment by Xenopathic** [comment]
@MorrisJobke I used `fopen()` and related functions so we only need to open the file and check for access once - before we do any reading or replacing of the contents. The original code used `file_get_contents()` and `file_put_contents()`.

**Comment by Xenopathic** [comment]
Unit tests! There was also a bug with checking the upper limit on the passed upload
size. PHP does funny things with integer vs float comparisons, so our
check didn't work. Now the check is much simpler, and ensures the value
is sane.

**Comment by Xenopathic** [comment]
@DeepDiver1975 @MorrisJobke Re-review please?

**Comment by Xenopathic** [comment]
@MorrisJobke Do you still think this should be using `file_get_contents()` and `file_put_contents()` instead of `fopen()`?

**Comment by MorrisJobke** [comment]
> @MorrisJobke Do you still think this should be using file_get_contents() and file_put_contents() instead of fopen()?

I'm fine with the way it is now.

**Comment by Xenopathic** [comment]
@DeepDiver1975 @MorrisJobke please review

**Comment by MorrisJobke** [comment]
Tested and works :+1: (also with changed permissions)

**Pull request merged**